### DISCUSSION TITLE
Can I remove x-middleware-rewrite http header from the response headers when using NextResponse.rewrite?

### DISCUSSION THREAD

**Discussion by Shelco** [vercel/next.js#58366] [Help]
### Summary

Hi,

We´re using NextResponse.rewrite in our middleware for passing certain requests to an internal api. The api in question has safety measurements set up, and is only reachable by whitelisted networks, but we would still prefer it if the actual internal url was not displayed to external users. However, when performing such a rewrite, the http header "x-middleware-rewrite" is set on the response and contains the internal url. Is there any way to remove this header from the response?

Best regards,
Dan

**Comment by liamlows** [comment]
Would love some input on this issue as well, we use rewrites to not expose our internal apis to the internet and this feature just leaks the internal URI for those APIs which is not great.

**Comment by TryV** [comment]
Never understand the mind behind this ideas. why everything is locked and limited in Next.js! you can't remove headers. you can't access the request object in page components. what back-end framework is this!

**Comment by jayantbh** [comment]
I've been unable to find a related issue (open or closed) for this as well. Will need help.

**Comment by davidhu2000** [comment]
Took a look in the source code, i "think" the header is added here https://github.com/vercel/next.js/blob/6bebacb2202bae04eebb0d10b44c829909cc3d9a/packages/next/src/server/web/spec-extension/response.ts#L135-L144

doesn't look like there is a way in NextJS to disable. 

The one solution I found, if you set that exact header where the request is rewrote to, the nextJS header is overwritten.

the client appears to get whatever value you set. 

For example.

```py
response.headers["X-Middleware-Rewrite"] = "--"
```

**Reply by TryV** [reply]
It didn't work for me.

**Comment by dpmango** [comment]
+1 for the option to hide `X-Middleware-Rewrite` which ruins all anonymization purposes of using api proxy

When manually overwriting header, it becomes always pending connection
```js
middleware(req: NextRequest) {
    const response = NextResponse.rewrite(url)
    response.headers.set('X-Middleware-Rewrite', '')
    return response
}
```

**Comment by JamieS1211** [comment]
Just ran into this issue, have any of you found any work-arounds?

**Reply by TryV** [reply]
No, you can reverse proxy your app behind the Nginx and remove header using nginx rules.

example:
```
  location / {
    ...
    proxy_hide_header X-Middleware-Rewrite;
    ...
  }
```

I didn't tried myself though.

**Reply by JamieS1211** [reply]
I settled on just returning the result of a fetch. Seems to work fine but dunno if this will have any unintended consequences down the line

**Comment by JamieS1211** [comment]
I revisited this issue and managed to solve it.

In `src/server/lib/router-utils/resolve-routes.ts` line 587 the early return seems to be the rewrite functionality.

https://github.com/vercel/next.js/blob/6bebacb2202bae04eebb0d10b44c829909cc3d9a/packages/next/src/server/lib/router-utils/resolve-routes.ts#L578-L592

I then was able to delete the unwanted header by adding the following between lines 586 and 587:

```js
delete resHeaders["x-middleware-rewrite"]
```

Having made this change I then used patch-package to ensure this change is always present - https://www.npmjs.com/package/patch-package

**Comment by vitminakov** [comment]
I encountered the same issue. Could you please suggest a possible solution? Would hiding this header on the Nginx side help?

**Reply by JamieS1211** [reply]
Give my solution a try, only takes 5 mins and works well for me

**Comment by ludvigeriksson** [comment]
Is there still no official way of doing this? The documentation states that "Rewrites act as a URL proxy and mask the destination path": https://nextjs.org/docs/app/api-reference/config/next-config-js/rewrites

**Comment by tobinbc** [comment]
I have "solved" with a Cloudflare rule to remove the header. It appears that Next actually does the rewritten call to whatever value you set the header to, AFTER the NextResponse is called. So if it's modified in any way after then the rewrite does not work:

```
const rewriteResponse = NextResponse.rewrite(new URL(destination));
rewriteResponse.headers.set("x-middleware-rewrite",'https://bbc.com'); // Next will try to return the bbc.com html to your request
rewriteResponse.headers.set("x-middleware-rewrite",''); // Next will not return the request as the header is deleted
```
